### ISSUE TITLE
Feature: Support multiple searches

### ISSUE THREAD

**Issue by KalitaAlexey** [microsoft/vscode#16488]
Hi everybody.

Currently VSCode supports only one active search: I cannot look for multiple things at a time.
I often have need in this feature.
I can implement it myself if someone will guide me.
I suggest to add new tab "Searches", which looks like this:
![collapsed_searches](https://cloud.githubusercontent.com/assets/11215191/20879396/a9d6fd40-bae4-11e6-9237-2ee4759a96c4.png)
And it expands to:
![expanded_searches](https://cloud.githubusercontent.com/assets/11215191/20879407/b8e989c4-bae4-11e6-8515-7344506cc9ea.png)
Unfortunately, I haven't made a close button. Imagine there is a close button to the right of a search entry.
When a user starts a search, then an entry in the searches is created.
When a user clicks on a search entry, then the entry's search becomes active.
When a user clicks on the close button of a search entry, then the entry is deleted.

I am waiting for your opinions to this feature.

**Comment by sandy081** [comment]
@KalitaAlexey Very much appreciated with your thoughts and ideas. I see that your base requirement is that you wanted to look multiple things at a time. Can you please explain in little more details how do you want to see above results at the end. Possibly with an example would be good.

Thanks

**Comment by KalitaAlexey** [comment]
Hi @sandy081,
Only one search is active at a time.
Results of the search is shown under the configuration of a search.

**Comment by KalitaAlexey** [comment]
Hi @sandy081,
I try to implement it myself.

**Comment by sandy081** [comment]
@KalitaAlexey We are always welcome to receive PRs for VS Code. But I would suggest before starting, it would be good to discuss in detail. 

- What is the problem?
- What is the proposed solution?
- Are there any other current ways to overcome the problem

Hence, I asked to know how the end result looks like and is it going to fix the problem? It would be helpful if you can explain me with an example.

Thanks

**Comment by KalitaAlexey** [comment]
Hi @sandy081,
I am downloading source code. I am changing the source code just to prepare an example.
I am sending pictures then.

**Comment by KalitaAlexey** [comment]
Hi @sandy081 and everyone,

I have thought about this feature and have prepared some screenshots explaining the design.
I'm sorry for the close button on the screenshots below.
Let's assume we open a project.
We navigate to the "Search" tab.
It is empty:
![empty_search](https://cloud.githubusercontent.com/assets/11215191/21010470/a2c2644e-bd65-11e6-9546-1bf6f1c677be.png)
We enter *A* to the "Search" field and press the Enter button.
It starts a new search.
When the search is ended, the "Searches" panel appears:
![a_is_added](https://cloud.githubusercontent.com/assets/11215191/21010741/31daa032-bd67-11e6-9f20-49ca61907c02.png)
We enter *B* to the "Search" field and press the Enter button again.
It starts a new search.
When the search is ended, in the "Searches" panel the new search appears:
![b_is_active_search](https://cloud.githubusercontent.com/assets/11215191/21010744/3759888e-bd67-11e6-8485-32d17469385f.png)
I click on the "A" search in the "Searches" panel.
The "A" search becomes active:
![a_is_active_search](https://cloud.githubusercontent.com/assets/11215191/21010748/3b0ee10e-bd67-11e6-9793-a629e93633ec.png)

If the close button of a search is clicked, then the associated search is removed.
If all searches are removed, the "Searches" panel is hidden.

# Unresolved questions
## Should a new search be added to the "Searches" panel either when the search started or ended?
If it would be added when it is started, then we should think about whether a running search may be cancelled.

## Is an active search customizable?
If it would be, then there should be a way to start a new search.

Regards,
Alexey

**Comment by sandy081** [comment]
@KalitaAlexey Thanks for the details. I am wondering why not using the following flow with current design (Single Search)

-> Search for `value1` you see results
-> Search for `value2` you see results
-> If you want to see previous results, navigate history in Search input using `alt+up` and `enter` and you see results.
-> if you want to see results matching `value1` or `value2` at a time why not use Reg ex  `(value1)|(value2)

I am not getting the use case that needs to show results for two search terms at a time. May be I am not sure, but I have not seen in any editor that shows multiple search results at a time. 

Sorry, but I am trying to get the need to build this functionality and how everybody can gain by this.

Thanks

**Comment by KalitaAlexey** [comment]
@sandy081,
I didn't know about `alt+up`, but anyway it doesn't preserve search results. I must start search again.
Think about it if a project very very big.

**Comment by leup** [comment]
Being able to keep search results would be very nice. Sometimes it's useful to be able to switch between search results without relaunching the search everytime.
So, may be a checkbox "Open results in a new pane", or something.
Netbeans, for example, keep results in different panes.

![image](https://cloud.githubusercontent.com/assets/431119/22327659/a13c38f6-e3b8-11e6-9062-33d3bcb383c7.png)

Here, for instance, I searched for "grdf" and "test", and each result set is kept in a different pane.

**Comment by theclapp** [comment]
You can use regular expressions to search for multiple items at once, though it's obviously still all the same single search.  E.g. from your example above, "A|B" would search for A and B at the same time.  (Make sure you enable regular expressions for that search; on a Mac, Opt-Cmd-R.)

**Comment by KalitaAlexey** [comment]
@theclapp,
I'd rather open a new window and search there.

**Comment by leup** [comment]
@KalitaAlexey 👍 
@theclapp I agree that it could be a solution for some simple searches, but this is not what I am looking for.

**Comment by theclapp** [comment]
Use parentheses: (ugly-complex-expression)|(some-other-expression)

Don't get me wrong: You can ask for this (or implement it!), that's fine with me.  I'm not saying you shouldn't.  I'm just saying there are alternatives or workarounds out there.

If someone does do this, might I suggest highlighting the different finds in different colors.

**Comment by KalitaAlexey** [comment]
@theclapp,
I know about alternatives, but it isn't convenient.
I got your point.

**Comment by jjinux** [comment]
This is an absolute must for me and a few of my co-workers. We have a very large codebase. Currently, I use IntelliJ. I'm often involved in very large refactors. I might do a search for something that I need to refactor. I work my way down the list slowly. I might have that search tab open for a day and a half. If I need to search for something else, I can search in a new tab without losing my place in my refactoring effort.

Switching between search terms is not a viable solution to this problem since I lose my place in the search results every time I have to do a new search.

This is a feature that NetBeans, IntelliJ, and even Sublime Text (via a plugin) have.

If you'd like to see how IntelliJ did the UI:

 * Do a global search.
 * Do another global search, but before you start it, click Result options :: Open in new tab.

By default, the "Open in new tab" is unchecked so as not to confuse the newbies. However, more advanced users such as myself have it checked all the time.

**Comment by spencer-schumann** [comment]
I agree that this is an important feature. Often when analyzing the results of a search it's helpful to run additional exploratory searches without losing the context on the original search. One helpful feature that search currently has is a close button to remove individual files from the search to prune out irrelevant results. If you've been analyzing and pruning a search and then start a new search, you currently lose all of the pruning you've done.

**Comment by balajitrv** [comment]
@sandy081 , Although i have nothing extra to add, VS2015, provides this option via ctrl+shift+f, append results or list results in results 2 window. This would be a good addition to VSCode if you can accommodate the request.

**Comment by KalitaAlexey** [comment]
But please don't limit count of searches to 2.
Sometimes you need much more than that.

**Comment by asad-abbasi** [comment]
I like VS Code very much and try to use it as much as possible. However multiple searches is one feature for which I am forced to get help from some other editor. I would suggest taking a look at Notepad++ before designing the feature for VS Code.

**Comment by hvasishth** [comment]
+1 to multiple search windows. It is very helpful when you want to do follow up searches to understand the context of something you found in the first search. I often have to fire up Idea just for this specific thing. Would like to be able to move to VSCode for everything.

**Comment by jjinux** [comment]
One nice workaround to handle this problem is to do:

    ag --group search_term directory | code - &

You can even do it from the terminal within VS Code. This will pop open a temporary file with the search results, and you can edit or delete them as you go through them.

**Comment by jjinux** [comment]
I blogged about this problem including which editors get it right and how to workaround the fact that VS Code gets it wrong. See: http://www.jjinux.com/2018/01/doing-multiple-searches-in-vs-code.html

**Comment by AndrewMorsillo** [comment]
I agree that this is a very important feature! I use (and love) VSCode for everything _except_ refactoring on my large code base. For that I have to open up Jetbrains WebStorm because it supports many searches at once. I frequently run into situations where I need to "keep my place" in one (or more) searches while doing another search to find other related references. This is the one feature missing from vscode that would make it absolutely perfect IMO!

**Comment by leocck** [comment]
I also upvote for this feature.

We are currently refactoring some big code, and it's very important for us to keep track of more than one search at the same time (to dive one or two more levels in).
Other editors have this functionality, allowing search results to be opened as a separate tab, as [Visual Studio](http://dailydotnettips.com/2016/01/30/did-you-knowthere-are-multiple-find-results-window-in-visual-studio/) and [Eclipse](http://billcomer.blogspot.nl/2012/01/multiple-search-windows-in-eclipse.html) do.

Any chance this could be implemented soon?

**Comment by efc** [comment]
Another upvote.

I work to construct a proper search, which includes decisions about the search term, the modifications of the search (case sensitivity and such), the appropriate replacements, and the files to include/exclude. It is really a package deal, all of it together is what represents a search, not any one field alone.

Right now I have to alt-arrow in at least three fields to sync up a package search that was all in sync at the time I last used it. And even then, the replace field and the search attributes (case, grep, etc) are left out and not recalled from history.

I feel less strongly about memorizing the search results, though that would certainly do no harm.

My model for this is BBEdit on the Mac, which keeps both an automatic history of searches executed and allows me to save named searches ("patterns" as it calls them). I would imagine a much simpler interface than the separate tabs or window (though, again, no harm in them, that would make me happy too). I would simply add a new mini-icon at the top of the search sidebar that pops up into a menu of saved searches with an option to save the current search.

**Comment by midGitub** [comment]
sometimes, show two different search result at the same time and at two split view  are more intuitive .
since the new version of vscode support splitview of terminal .
i hope there are some plan such as splitview of the search panel.

**Comment by hsaih** [comment]
Any update about this feature?

**Comment by jcory** [comment]
Agreed that this would help significantly with large refactors

**Comment by KristjanTammekivi** [comment]
This is really something I miss from Sublime Text

**Comment by AndrewMorsillo** [comment]
Any update about this feature? This is the number 1 missing thing in vscode for me.

**Comment by dev6rodn** [comment]
Here's my upvote. The task of following long, convoluted trails through a large, complicated code bases would be easier if I had the ability to maintain multiple independent searches. 

When I'm trying to discover the root cause of an issue or develop an understanding of how a feature works, it's never a one-search exercise. I need the freedom to explore the code, weed out less interesting results and continue the exploration exercise (in a separate  search tab) based on whatever I discover. Sometimes those search "side roads" turn out to be dead ends but I almost never want to sweep away my back trail as I explore because every search result represents a potential new avenue to snoop around in. It's a process and the job is hard. Multiple searches would make it easier.

Thanks for an awesome product! Here's to keeping up the great work.

**Comment by leomeneguzzi** [comment]
What is the situation for this feature? Someone are making this? I have some ideas about it and if nobody is doing this I maybe can try.

**Comment by midGitub** [comment]
@leomeneguzzi  
when looking for all the place using keyword A and   all the place using keyword B. compare their usage way and used frequency.

**Comment by Daijobou** [comment]
I'm not sure what this feature means. Is this a single search with more search patterns and result is a **single list of results** or is the result that each search pattern have a own independent list of results and I can switch between this list of results?

**Comment by dev6rodn** [comment]
I was envisioning being able to switch between multiple independent searches, each with its own pattern and results.

**Comment by jovere** [comment]
For an example of a good way to do search results, take a look at the following screenshot from Notepad++.  This is showing two searches, with the top one fully collapsed and the one below having some of them opened.  The complete result history stays until the session is ended or individual sections/lines are removed (pretty much like VS code does now, except it only shows the results of one search).

![image](https://user-images.githubusercontent.com/10420943/44663221-85a34c80-a9d5-11e8-9745-93d7e431159c.png)

**Comment by lovelyjuice** [comment]
I want to use vscode to audit code, and show my friends that it was better than sublime. But unfortunately it lack only one the most important feature.  If it can do it over time, please remind me, thanks!

**Comment by lovelyjuice** [comment]
I want to use vscode to learn code audit, and show my friends that it is better than sublime. But unfortunately it lacks only one the most important feature. If it can do it over time, please remind me, thanks!

**Comment by adi-it** [comment]
I need per window search like the one in firefox or chrome.

**Comment by AndrewMorsillo** [comment]
vscode is awesome! IMO this is the _only_ major missing feature preventing it from being perfect =)

**Comment by waterdeng** [comment]
Almost two year since this issue was created. Is **multiple search** on the roadmap? I really need it.

**Comment by amituuush** [comment]
kind of a bummer this isn't supported yet

**Comment by brunofin** [comment]
It seems I reported a duplicate feature request, so I will add here my take on the feature: use tabs?

Here is a simple mockup to illustrate the idea:

![vscode-search-on-new-tab](https://user-images.githubusercontent.com/8077237/48658910-c85f4880-ea49-11e8-903b-029408d9255c.png)

**Comment by Stuey2** [comment]
My two cents:

VSCode has one of the most powerful search functionalities on the market in a number of ways, I believe. 
a) It allows solid regex, etc. searches with white and black lists for files. and .gitignore, etc.
b) It has tracks history all over the place for previous searches
c) It allows one to dismiss irrelevant search results (awesome!)

W.R.T (c) especially, but also for say a complex regex search. There is a need to allow for some sort of search persistence. Whether that is via multi-search tabs or saving and retrieving named searches. This is so, because if one goes to great care to craft a search and then filter it through the dismiss functionality, then there's an appreciable amount of work into just performing that search. 

I suspect that this is also not the end of the road for search either though. I think it's self-evident these days that a deep search capability is highly attractive and hugely aids productivity and also the more one relies on search the more one needs it.

Having said all that I think getting the GUI conceptually right is very important. Merely having tabs isn't going to scale for any number of searches. Also having loads of randomly named searches might end up in a mess. There needs to be a good way of bringing meaning to the 'named / multi search' capability. (I think that might be the hold up here)

If this could be implemented nicely though it would be fantastic!

**Comment by p-bakker** [comment]
Maybe missed it but no-one mentioned Eclipse yet. The list of features mentioned in the above comment that make VS Code search so powerful are all supported in Eclipse as well.

The concept of multiple Searches is also possible in Eclipse: you search on something and if you want to keep the result, you pin the result, which effectively means that a next search will open in a new (results) tab. Difference with VS Code is that in Eclipse the Ui for doing the search is separate form the UI showing the results.

This ability to have multiple search results open (thorugh pinning) is hugely beneficial in exploring a (large) codebase and/or working on large refactors impacting many, many files.

Problem seems to be that the VS Code team is not eager to move towards a tabbed interface for non-editors, see https://github.com/Microsoft/vscode/issues/10546

**Comment by rnorton1** [comment]
I would also like to upvote this feature. Coming from Eclipse, the ability to be able to flick between prior searches is the one feature glaringly missing for me.

As this request has been open for some time, I think we would appreciate some feedback on the likelihood of this feature getting implemented, or closing if the answer is definitively no.

Thanks for your time.

**Comment by AndrewMorsillo** [comment]
This is still the only reason I keep Webstorm installed on my machine. I simply can't do some complex refactor tasks with VSCode because of the lack of multiple searches. The feature in Webstorm is fairly simple... Each search shows up as another tab with the title of the tab being the search term. I would see this working well in vscode as well. For example the tabs in the default bottom window would be "problems, output, debug console, terminal, search results". When search results is open there are sub tabs (or some other selector) to switch between the active searches.

**Comment by robba86** [comment]
Just do the feature already? :)

**Comment by AndrewMorsillo** [comment]
I certainly would if I had the time and the expertise with the vscode code base. Unfortunately I don't have either of those so I just wanted to contribute my thoughts on how the design might work and what use cases this feature request would solve.

**Comment by jimmybrawn** [comment]
lets start 2019 with this feature ye? 🤣

**Comment by dbechrd** [comment]
This would save me an insane amount of time on a daily basis.. the problem is compounded by the fact that you can't open VSCode twice in the same directory (it just focuses the already open window) so it can't even be worked around by having two instances of VSCode running.

Having to delete my search during a refactor to look for something else, then recreate it manually and re-prune the list of results is tedious and unnecessary. :(

**Comment by hyperwired** [comment]
It is disappointing that this is still open after 2 years. It is a really important feature for working with and understanding large codebases that you may not be overly familiar with. At the risk of potentially repeating some of the previous discussion, here are a couple of UX options that can be adapted, and have been proven in other products:

- Sublime Text | Find in Files: generate a new "find results" tab with clickable search results, possibly including context lines.
- Visual Studio | Find in Files (with append results option checked): Has a "Find Results" output pane where multiple sets of search results get added where each search is a collapsible entry that gets added the the pane. 
  - This may be a preferable option if there is concern over adding tabs.
  - The output could be in a new "find results" pane that sits next to output/debug console/terminal), 
  - Stretch goal 1: 2 or more "find results" panes with options for which one the search results are appended to (e.g. a search options radio or drop-down box setting, similar to VS)
  - Stretch goal 2: Add context lines to search results (configurable)
  - Stretch goal 3: Add a similar pane for "Find Reference Results" panel, that is also appendable, since "find references" results have the same limitations as Results

Visual Studio UI example for those not familiar:

VS Find in Files dialog:
![image](https://user-images.githubusercontent.com/1922244/51288418-4cd14600-1a50-11e9-9263-03b48c912df5.png)

VS Find Results output panel (note that each result line is clickable, but all the result text is copyable as plain text, including ctrl+A, which can be handly):
![image](https://user-images.githubusercontent.com/1922244/51288401-31fed180-1a50-11e9-91d6-59121fa80633.png)

Having either the sublime or VS style options to start with should not preclude changing the interaction paradigm later if it is really necessary, but both are battle tested in industry in very popular editors, and either option (or a different one) is better than none at all... Is the UX really so immutable that it must be perfectly designed up front and cannot be tweaked or changed later?

I think adapting Visual Studio behavior could be a very good starting point for this feature.

**Comment by lovelyjuice** [comment]
> [Investigate showing search results not only in the side bar or a panel, but also in an editor. This allows us to show additional context information for each match.](https://github.com/Microsoft/vscode/wiki/Roadmap#workbench)

It seems to already on the Roadmap now.

**Comment by joelewis** [comment]
Team,
This request by now has enough use-case validation to justify existence. Is this feature in the roadmap? If yes, any UI design ideas floating around? 

The request is pretty dated (3 years now). I'd be happy to jump-in and issue a PR if need be.

**Comment by RyanEwen** [comment]
I've been running into this problem more and more in time as I need to search unfamiliar code for something and then need to start a second search based on one of the results from the first, but I don't want to lose my context/the results of the first search in order to perform the 2nd. 

So +1 to the ability to have multiple searches, whether they be nested in the search view or tabs within the search view or opened as editors or whatever the case may be.

**Comment by Avivhdr** [comment]
Looking forward for this feature

**Comment by EricG-Personal** [comment]
I too would like to see this feature implemented. For an example of multiple search functionality that works extremely well, see Apple's Xcode application.

**Comment by creatyvtype** [comment]
Would like this implemented as well!

**Comment by NickGrease** [comment]
![Search](https://user-images.githubusercontent.com/854582/55289249-ecd11700-5391-11e9-9384-614bb8799679.png)

I hope the image is clear  - I'm constantly searching documentation and having those searches available without having to re-search every time would be a real boon.

**Comment by dpope-atlassian** [comment]
We have a large JS codebase that uses lots of dynamic-typed features (e.g. calculate a function name to call by concatenating string values :sigh:). When working with such a codebase you have to search a **lot**, unlike with statically typed languages like TS.

As someone else pointed out, in a system like this you'll be constantly trying to figure out how something works, using search to find things of interest, and while you're doing that you'll need to do another search. The amount of mental context required is absurd, honestly. So when you have to destroy your "overall" context (just to reuse the search input widget), it destroys the flow and forces the use of external tools if you want to keep context.

I need my tools to multiply my limited brainpower, and the absence of this seemingly-obscure but actually-critically-important-for-this-sadly-not-uncommon-use-case feature is a significant hindrance.

**Comment by joejordan** [comment]
Agree with @spencer-schumann on the need to save pruned searches. Like many others commenting, I am working on a large code base, and I have a search open at the moment that if I happen to lose, will require a significant amount of work to restore. Multiple tabbed searches with memory of pruned results are critically important features.

**Comment by ppggff** [comment]
Before this feature is implemented, vscode's 'task' can be used to simulate it.

**Comment by neakor** [comment]
Would love to have this please!

**Comment by amarcus-adobe** [comment]
I just ran into a need for this feature, so please consider this use-case a 👍for implementation:

I am refactoring a large React-based codebase, and searching for all instances of a top-level prop.  I love using the search result's "X" to dismiss instances of the variable that I have already checked, and have been doing so as I move through the code.

Now, I have found an instance of the variable that needs refactoring within a component class.  In order to do so, I need to search the codebase for instance of this component to ensure that its props are updated accordingly.

I'm stuck!  If I search for instances of the component, I will lose my progress in dismissing the original search results.

@NickGrease has a good UI treatment for my use case.  Maintaining multiple search results panels - particularly with their dismissal status - would be immensely helpful here.

**Comment by steelyvoid** [comment]
An uptick from me for this.  I regularly need to nest searches and its a pain, having gone down the search stack, to have to find an example of the string I was originally searching for and search again.

**Comment by LACoderDeBH** [comment]
Same here, this would be incredibly useful in refactoring code that contains instances of my type defs in a gql query.

**Comment by thatandyrose** [comment]
This is a killer feature and currently the only viable workaround is to open ANOTHER VSCode instance, which you'll agree is a bit unnecessary. Let me give you guys a concrete example.

I have a large Rails project where I want to first find all occurrences of a certain css class in my view templates.

The results yields all the partial views where that css class is referenced.

Now with the list of templates always visible in the results list, I now want to search, one by one, where each template is referenced.

In other words, one search provides a list. And the second search allows me to search for each items on that first list.

This is just ONE example of many. The idea of going back, searching again etc etc. Yes it's doable, but if the point of this is to make a really usable code editor then we shouldn't make do with a really horrible user experience of having only one search.

Thoughts oh VSCode gods? 😉

**Comment by jjinux** [comment]
I wrote a nice tweet about this issue in order to try to get it some attention. Please retweet and favorite it if you're into that sort of thing: https://twitter.com/jjinux/status/1153808934966140928

**Comment by mario206** [comment]
The feature is quite necessary to analyze a log file.
See how it works in Nodepad++
![image](https://user-images.githubusercontent.com/5844092/62356108-eb337d80-b542-11e9-8c0c-a87520a3f8c8.png)

**Comment by bogdan-coanda** [comment]
Same story, I need this feature for refactoring large code bases...

**Comment by ashish-r** [comment]
Feature needed for refactoring large code base.

**Comment by jaroslawjanas** [comment]
Just get it done.

**Comment by dzchoi** [comment]
Need this, since the Global Search can be thought of as a source code analyzing tool, not just as a find-and-forget tool.

**Comment by DOHere** [comment]
search function needs to improve

**Comment by akashagarwal7** [comment]
Need this big time

**Comment by bildungsroman** [comment]
Does anyone know of an extension that does this? I really need this functionality as I work in multiple repos daily and often need see the search results of one repo in order to search through another repo.

**Comment by aramando** [comment]
The obvious workaround to this requirement, opening a new Code instance, doesn't work because Code doesn't let you open the same folder in multiple windows. Please enable multiple searches so I can do a new search without losing my edited results from the last search!

**Comment by Duncan3142** [comment]
Does anyone know if the current architecture of vscode search is blocking this enhancement? Would love to see this implemented, just curious to know if there's a larger technical blocker that needs addressing first, and if it's something the community could work on...

**Comment by gel** [comment]
This is mandatory, how can we get this prioritized?

**Comment by RenjingYin** [comment]
Encountered same use case, I downloaded another editor for simple solution, but would love to see this in VScode

**Comment by JacksonKearl** [comment]
Hey all, this sprint I'll be working on a somewhat related feature, ["Search Tabs"](https://github.com/microsoft/vscode/issues/23931), where search results can be copied over to a tab while preserving full "go-to-result" support. You could have an arbitrarily large number of these "Search Tabs", but (at least for the initial implementation), the query will not be modifiable once the results have been dumped into a tab. 

I believe this would solve many of the pain points listed in this thread, and I'm interested to hear if anyone here has a use case that would not be satisfied with that model.

**Comment by aramando** [comment]
@JacksonKearl That sounds perfect, especially if the search tabs will be formatted in the same way as the search results, i.e. grouped by source file and with the option to permanently remove items from the list once they have been assessed to be either irrelevant or dealt with.

**Comment by JacksonKearl** [comment]
@aramando The search tabs will be text editors (a. la. Sublime), so removing results is deleting text.

**Comment by AndrewMorsillo** [comment]
@JacksonKearl Awesome! Glad to see movement on this. My opinion is that the search panel should be done away with altogether. Search results should always open in a separate persisted tab. If I think about all the searches I've done recently not a single one of them made sense as a temporary set of results in the side panel. Nearly all global searches I do lead to subsequent searches based on the original result set which is currently impossible to do while keeping the original results around. This lackluster search function is _by far_ my biggest pain point with VSCode.

**Comment by JacksonKearl** [comment]
@AndrewMorsillo This is just the MVP to start to get an idea of what the Search Editor flow is like, many people have similarly requested to skip the search viewlet/panel entirely and we may in the future make the overall search position configuration `'viewlet' | 'panel' | 'editor'`. #23931 has more discussion on the matter.

**Comment by Maxxee** [comment]
(Sorry for "invading" this thread - well, it's not really invading, it's about the search feature after all and all other threads are closed.)

I really appreciate that you made the search feature remember the search phrases by pressing up/down or alt+up/alt+down.
This was really needed.
Now excuse me, but what is so difficult about adding something visual to the **Visual** Studio Code like two arrow keys, so that you have an expandable list of all the previous search phrases like in any other editor?

**Comment by fxdave** [comment]
@JacksonKearl Having search tabs is a nice to have feature, but I would step a bit further. On every search tabs, there could be a "-" and a "+" button to be able to subtract and add other search tab results to(from) the actual one.

On search tabs:
By clicking "-" a modal could show up to select a search tab for subtraction.
By clicking "+" a modal could show up to select a search tab for addition.

On the selection modal:
By pressing "enter" it would modify the currently active search tab.
By pressing "shift+enter" it would create a new search tab.

Addition and subtraction should operate between filenames, instead of between the resulted lines since lines can be subtracted out by an in-tab search.

**Comment by JacksonKearl** [comment]
A new Search Editor implementation is out for testing, which enables having multiple active searches:
![image](https://user-images.githubusercontent.com/8586769/72850220-80dfd800-3c5d-11ea-892d-2b4285f60006.png)

You can try this out with the `Open new Search Editor` command, which you can map to any keybinding you'd like in the Keybindings Editor.

Let me know how this works for your multiple searches use cases.

More details at #88154.

**Comment by JacksonKearl** [comment]
As mentioned earlier, we are supporting multiple searches through search editors, which have been available in preview for a couple releases and will be out of preview in the February release.

**Comment by NickGrease** [comment]
I appreciate the effort to improve searches, but this is not what I had hoped for. I have more than enough tabs open as it is, adding a half dozen or more tabs is ridiculous. The side bar is where searches belong (for me, my opinion); horizontal workspace is at a premium, and I hope someone over there can appreciate other peoples workflow and viewing restraints.

**Comment by JacksonKearl** [comment]
I appreciate your feedback and understand that this isn't exactly what was originally requested, but at the end of the day we aren't in a position where we can dedicate resources to developing many different competing search models. With that said, I would be interested in improving one of the current search models to better suit your needs. 

> the tab doesn't even show what files/folders/disks are being searched

Could you explain what you mean by this? It uses the same query editor widget as the search viewlet. 

> I don't really enjoy the idea of scrolling through 1000's of lines of search results to see if a match was in a specific file.

Wouldn't you be doing this in the viewlet as well?

**Comment by jaroslawjanas** [comment]
What I rly end up needing all the time is
1. Nested searches
2. Search history/selection
3. More advanced searches
4. Better documentation on how to actually search stuff because the one you have at the moment is incredibly confusing (at least for me, maybe I'm stupid)

**Comment by roblourens** [comment]
@jaroslawjanas I would like to hear more about all of those points. Please open a new issue

1. Honestly I'm not sure what you mean by this
2. You can navigate history with the up/down arrow keys, please tell me more if it's not what you're looking for
3. Advanced in what way?
4. And I'd like to improve the documentation, please tell me what you are struggling with

**Comment by JacksonKearl** [comment]
> Nested searches

Do you mean searching for a string within the results of a different search? This is possible with search editors.

**Comment by JacksonKearl** [comment]
@roblourens 

I think 2. refers to a global history of the full search configuration, so queries get associated with flags and includes/etc. This could be especially useful now that we have two separate ways to run workspace searches.

**Comment by jaroslawjanas** [comment]
1 & 4. I just did a **quick** google search of (search editors didn't find anything in the top searches)
2. I mean a search history similar to history of actions in Photoshop, where you get a table with all of your actions listed in order, so something similar for vs code would be nice, rather than jumping from search to search till you find the one you wanted.

3. Good question, I don't know why but I find the feature lacking in features (no pun intended), I mean it shouldn't be overdeveloped either but maybe some more UI or something that would make complex searches easier to write. Maybe an option to have a preview based on sample data and then you can just swap your actual search params to do the actual search? That'd be pretty handy.

4. Usages, cases, examples of difficult searches.

**Comment by roblourens** [comment]
Please, let's discuss in a new issue

**Comment by NickGrease** [comment]
@JacksonKearl 

> > the tab doesn't even show what files/folders/disks are being searched
> 
> Could you explain what you mean by this? It uses the same query editor widget as the search viewlet.
![ImageA](https://user-images.githubusercontent.com/854582/75599802-6a0b7d00-5a76-11ea-8109-ada090d1d15a.png)

![ImageB](https://user-images.githubusercontent.com/854582/75599804-6aa41380-5a76-11ea-923d-6f23b0311e04.png)

With multiple tabs open, figuring out which tab was which search, is a pain compared to the sidebar view.

> > I don't really enjoy the idea of scrolling through 1000's of lines of search results to see if a match was in a specific file.
> 
> Wouldn't you be doing this in the viewlet as well?

![ImageC](https://user-images.githubusercontent.com/854582/75599803-6a0b7d00-5a76-11ea-925f-cd5138ad674d.png)

Using your treatment/implementation, the search for "Test" would mean scrolling past 64 results in Scripts, 24 in documentation, to finally find the 4 in research. Using the image above, it takes a split second.

When I'm searching hierarchies of folders (that I'm not always familiar with) for bits of api code/functions - I *need* to see at a glance what folders the results are in. If I search for a function that I expect starts with the word "Load", seeing a list of folders that have matches means I can guess what folder I'm likely looking for, for example,if the results show a dozen folders, one is "ShipAI", one is, ShipCrewAPI" one is "PlanetAPI" - if I was looking for how to "load crew" - I'd know instantly what folder to open results in.

I hope that makes sense; and at the end of the day I've been managing for well over a year now. You do you, and I'll make do :)

**Comment by frumbert** [comment]
> @jaroslawjanas
> 
> > 1. Nested searches
> 
> Do you mean searching for a string within the results of a different search? This is possible with search editors.

I'd like to search for a term to get a list of the files that match that value. THEN with the list of the files, search only those files for a different term. Currently the process is that I click the file in the search result, the press ctrl-f on the file view and search my second term inside the file, stepping through all the matches. It would be far easier to have a "Search within results" option.

What are search editors?

**Comment by JacksonKearl** [comment]
@frumbert 

> what are search editors?

See the release notes: https://github.com/microsoft/vscode-docs/blob/vnext/release-notes/v1_43.md#search-editors

![Search Editor
 UI](https://github.com/microsoft/vscode-docs/blob/vnext/release-notes/images/1_43/search-editor-overview.png)
